Formula/elan-init.rb

```diff
@@ -0,0 +1,50 @@
+class ElanInit < Formula
+  desc "Lean Theorem Prover installer and version manager"
+  homepage "https://leanprover-community.github.io/"
+  url "https://github.com/leanprover/elan/archive/v1.0.3.tar.gz"
+  sha256 "fb3ddd8915a0694ead0f3a51fdf8b0a5540f983f44eee0e757339244c522b8ee"
+  license "Apache-2.0"
+  head "https://github.com/leanprover/elan.git"
+
+  depends_on "rust" => :build
+  depends_on "gmp"
+
+  # elan-init will run on arm64 Macs, but will fetch Leans that are x86_64.
+  depends_on arch: :x86_64
+
+  def install
+    ENV["RELEASE_TARGET_NAME"] = "homebrew-build"
+
+    system "cargo", "install", "--features", "no-self-update", *std_cargo_args
```

Comment:
The way I understand this, `elan-init` will install a bunch of binaries (including `elan`) into `~/.elan/bin` (unless the user overrides that location).

Just to double-check, the `no-self-update` feature deactivates updates for `elan-init` (we want this), but not any of the ones installed into `~/.elan/bin` (we don't want this), correct?

Reply:
> The way I understand this, `elan-init` will install a bunch of binaries (including `elan`) into `~/.elan/bin` (unless the user overrides that location).

Yep exactly.

> Just to double-check, the no-self-update feature deactivates updates for elan-init (we want this), but not any of the ones installed into ~/.elan/bin (we don't want this), correct?

Yep (well, technically it disables it for the `elan` binary as well I think) -- but not for `lean` and friends, those versions elan will happily upgrade. So basically Homebrew then manages elan's version, but that fixed elan version can manage any current past or future lean toolchains.

Reply:
Ok, but when `brew` updates `elan-init`, how do we also update the `elan` that lives in `~/.elan/bin`?

Reply:
Sorry I misspoke I believe -- which I just checked by arbitrarily bumping this back to 1.0.2 and then doing a `brew upgrade` to this version, and seems to match the below.

The things that end up in `~/.elan`, specifically in `~/.elan/toolchains/*/bin/` are Lean versions, so the `lean`, `leanpkg`, `leanchecker`, `leanc` and `leanmake` binaries for each specific version.

The `elan` binary, as well as symlinks to it for `lean` and `leanpkg` and the rest here which magically delegate to the right lean version, are just the ones that Homebrew will link to `brew --prefix/bin`, so there's no separate dangling `elan` that ends up in `~/.elan/bin/`.

Hopefully that makes sense now.

Reply:
Wait, so all these binaries that live in `~/.elan/bin` are all just copies of `elan-init`?

```
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 elan
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 lean
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 leanc
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 leanchecker
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 leanmake
.rwxr-xr-x 4.8M carlocab 24 Nov  2020 leanpkg
```

I guess that checks out:

```
❯ for f in ~/.elan/bin/*; do diff -q ~/.elan/bin/elan $f && echo "elan and $(basename $f) are the same"; done
elan and elan are the same
elan and lean are the same
elan and leanc are the same
elan and leanchecker are the same
elan and leanmake are the same
elan and leanpkg are the same
```

Kinda annoying that I have six different copies of the exact same thing though.

Reply:
That's from a previous run of `elan-init` that you installed in 2020 right, not from running this formula? If you instead install from this formula you don't get those? (So you'll get just symlinked versions the way I wrote things here).

That's what happens for me at least when using this formula (I don't get an `~/.elan/bin/`, and yeah symlinks instead of 6 hard copies)

Reply:
> That's from a previous run of `elan-init` that you installed in 2020 right, not from running this formula?

Yes, that's from an installation using the [instructions I linked to above](https://leanprover-community.github.io/install/macos.html) that I followed a while back. `ls -l` shows the targets of symlinks, so it would've shown in my previous comment if they were symlinks.